### PR TITLE
[GENX] Add `genx.matrix.dpas` (part 1)

### PR DESCRIPTION
**Pull request by whitneywhtsang**
The 'genx.matrix.dpas' operation is a matrix multiply-add operation as follows:

       D = C + A x B

       where
          D : MxN
          C : MxN
          A : MxK
          B : KxN

          M : repeat count ($rc)
          N : fixed execution size, must be 16 for PVC
          K : depth * OPS_PER_CHAN
              OPS_PER_CHAN
                 1 : for TF32
                 2 : for 16-bit precision(BF, HF)
                 4 : for 8-bit precision (FP8, UB, B)
                 8 : for less-then 8 bit precision (U4/S4, U2/S2).

              If depth is 8, K would be 8, 16, 32, or 64 (based on OPS_PER_CHAN).

### CODE REVIEW

File: mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp [hunk 1]
```diff
@@ -203,6 +203,33 @@ static llvm::Value *createAtomicRMW(llvm::IRBuilderBase &builder,
...
+                 LLVM::ModuleTranslation &moduleTranslation) {
+  llvm::Module *module = builder.GetInsertBlock()->getModule();
+  auto opTypes = op->getOperandTypes();
+  llvm::Function *fn = llvm::GenISAIntrinsic::getDeclaration(
+      module, llvm::GenISAIntrinsic::GenISA_dpas,
+      {moduleTranslation.convertType(op->getResultTypes()[0]),
+       moduleTranslation.convertType(opTypes[0]),
+       moduleTranslation.convertType(opTypes[1]),
+       moduleTranslation.convertType(opTypes[2])});
+  // FIXME: Attribute mismatch - Attribute 'noundef' does not apply to
```
Comment (whitneywhtsang) [inline mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp]: @pengtu Is the attribute mismatch due to llvm mismatch?

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td [hunk 2]
```diff
@@ -338,7 +338,64 @@ def GENX_AtomicRMWOp : GENX_Op<"atomic.rmw", [
...
+
+    $a, $b, $c, $d - matrix A, B, C, D, respectively
+    $pa, $pb - precision of matrix A and B resepectively
+    $sd - systolic depth
+    $rc - repect count
+    $isDpasw - is dpas wide instruction
+  }];
+
+  string llvmBuilder = [{
+    $d = createGenISADPAS(op, builder, moduleTranslation);
```
Comment (whitneywhtsang) [inline mlir/include/mlir/Dialect/LLVMIR/GENXOps.td:387]: I think passing op is cleaner than passing each argument. Will change `GENX_Matrix2DBlockLoadOp` to do the same, if no one disagrees.

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td [hunk 3]
```diff
@@ -338,7 +338,64 @@ def GENX_AtomicRMWOp : GENX_Op<"atomic.rmw", [
...
+  }];
+    
+  let assemblyFormat = [{
+    operands attr-dict `:` functional-type(operands, results)
```
Comment (whitneywhtsang) [inline mlir/include/mlir/Dialect/LLVMIR/GENXOps.td:391]: I think parsing/printing attribute with `attr-dict` is cleaner. Will change `GENX_Matrix2DBlockLoadOp` to do the same, if no one disagrees.

File: mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp [hunk 4]
```diff
@@ -15,6 +15,15 @@
 
 using namespace mlir;
 
+//===----------------------------------------------------------------------===//
+// genx.matrix.dpas
+//===----------------------------------------------------------------------===//
+
+LogicalResult GENX::MatrixDPASOp::verify() {
+  // TODO
```
Comment (whitneywhtsang) [inline mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp:23]: I would like to first ensure `genx.matrix.dpas` representation is good before working on the verification code.

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td [hunk 5]
```diff
@@ -338,7 +338,64 @@ def GENX_AtomicRMWOp : GENX_Op<"atomic.rmw", [
...
+    FixedVectorOf<[GENX_MatrixElemType]>:$a,
+    FixedVectorOf<[GENX_MatrixElemType]>:$b,
+    I32Attr:$pa,
+    I32Attr:$pb,
+    I32Attr:$sd,
+    I32Attr:$rc,
+    I1Attr:$isDpasw
+  )> {
+
+  let summary = "GEN matrix multiply-add";
```
Comment (etiotto) [inline mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: nit: GEN -> GENX

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td (hunk 2 above)
Comment (etiotto) [reply mlir/include/mlir/Dialect/LLVMIR/GENXOps.td:387]: I agree

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td (hunk 3 above)
Comment (etiotto) [reply mlir/include/mlir/Dialect/LLVMIR/GENXOps.td:391]: Agree

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td [hunk 6]
```diff
@@ -338,7 +338,64 @@ def GENX_AtomicRMWOp : GENX_Op<"atomic.rmw", [
...
+  Results<(outs FixedVectorOf<[GENX_MatrixElemType]>:$d)>,
+  Arguments<(ins 
+    FixedVectorOf<[GENX_MatrixElemType]>:$c,
+    FixedVectorOf<[GENX_MatrixElemType]>:$a,
+    FixedVectorOf<[GENX_MatrixElemType]>:$b,
+    I32Attr:$pa,
+    I32Attr:$pb,
+    I32Attr:$sd,
+    I32Attr:$rc,
+    I1Attr:$isDpasw
```
Comment (etiotto) [inline mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: I am wondering whether we should have 2 separate operations, one for the dpas instruction and one for the dpasw instruction (similar to NVIDIA which has the MMA and WMMA operations). That way the user will not have to set the `isDpasw` boolean. I think having 2 operation would be cleaner from the perspective of the user. Also the Intel assembly instruction are separate, we have one for dpas and one for dpasw.

In this PR I suggest to drop the `$isDpasw` argument given we just need the dpas instruction right now. Later we can decide on whether we want to add support for dpasw too.

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td [hunk 7]
```diff
@@ -338,7 +338,64 @@ def GENX_AtomicRMWOp : GENX_Op<"atomic.rmw", [
...
+       D = C + A x B
+
+       where
+          D : MxN
+          C : MxN
+          A : MxK
+          B : KxN
+
+          M : repeat count ($rc)
+          N : fixed execution size, either 8 or 16
```
Comment (etiotto) [inline mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: This has to be 16 for the PVC architecture and has to be 8 for previous GPU generations. So the question is  whether we want to support just PVC or not. My opinion is we make this mandatory to be 16. Propose to change this text to:
```
  N : fixed execution size, must be 16 for PVC

File: mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp (hunk 1 above)
Comment (etiotto) [reply mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp]: This is very likely due to the LLVM IR mismatch we discussed. The FIXME IMO can be addressed after peng work on IGC is done.

Comment (pengtu) [reply mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp]: Yes, the workaround is at https://github.com/pengtu/llvm-project/blob/genx/mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp#L228

File: mlir/test/Target/LLVMIR/genx.mlir [hunk 8]
```diff
@@ -135,8 +135,15 @@ llvm.func @genx.atomic.rmw(%ptr : !llvm.ptr<i32, 1>, %sptr : !llvm.ptr<i64, 3>,
   llvm.return
 }
 
+llvm.func @genx.dpas(%c : vector<8xi32>, %a : vector<8xi32>, %b : vector<8xi32>) {
+  // CHECK: %4 = call <8 x i32> @llvm.genx.GenISA.dpas.v8i32.v8i32.v8i32.v8i32(<8 x i32> %0, <8 x i32> %1, <8 x i32> %2, i32 32, i32 4, i32 1, i32 1, i1 false)
+  %0 = genx.matrix.dpas %c, %a, %b {pa=32:i32, pb=4:i32, sd=1:i32, rc=1:i32, isDpasw=false} : (vector<8xi32>, vector<8xi32>, vector<8xi32>) -> vector<8xi32>
```
Comment (etiotto) [inline mlir/test/Target/LLVMIR/genx.mlir]: I think the elem. type of %a and %b has to be i8 when C has elem. type i32. Anyway once the verification code for the operation is added illegal combination of elem. types and attribute values should be diagnosed.

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td (hunk 2 above)
Comment (pengtu) [reply mlir/include/mlir/Dialect/LLVMIR/GENXOps.td:387]: I agree.  Thanks.

File: mlir/test/Target/LLVMIR/genx.mlir (hunk 8 above)
Comment (pengtu) [reply mlir/test/Target/LLVMIR/genx.mlir]: It is probably better to use f32 type and PVC parameters for the test case to illustrate its usage: %c : vector<8x16xf32>, %a : vector<8x8xf32>, %b: vector<8x16xf32>, pa=32, pb=32, sd=8, rc=8

Comment (pengtu) [reply mlir/test/Target/LLVMIR/genx.mlir]: Maybe even %b: vector<16x8xf32> for the column major layout.

**Review by pengtu — APPROVED** [review]

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td (hunk 6 above)
Comment (pengtu) [reply mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: I agree.

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td (hunk 3 above)
Comment (pengtu) [reply mlir/include/mlir/Dialect/LLVMIR/GENXOps.td:391]: Agree

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td (hunk 6 above)
Comment (whitneywhtsang) [reply mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: Dropped `isDpasw`.

File: mlir/test/Target/LLVMIR/genx.mlir (hunk 8 above)
Comment (whitneywhtsang) [reply mlir/test/Target/LLVMIR/genx.mlir]: multi-dimensional vector is not supported in LLVM.

Comment (whitneywhtsang) [reply mlir/test/Target/LLVMIR/genx.mlir]: Let's keep this test case for now, having difficulty to understand what we should allowed on PVC. We can change it when we understand better.

**Review by etiotto — APPROVED** [review]

File: mlir/test/Target/LLVMIR/genx.mlir (hunk 8 above)
Comment (pengtu) [reply mlir/test/Target/LLVMIR/genx.mlir]: Okey.  We can probably linearize the 2D submatrix to a 1-D vector: %c: vector<128xf32>, %a: vector<64xf32>, %b<128xf32>. Let's do it later then.

Comment (whitneywhtsang) [reply mlir/test/Target/LLVMIR/genx.mlir]: From the verification code https://github.com/intel-innersource/drivers.gpu.unified/blob/comp/igc/Source/IGC/Compiler/Optimizer/OpenCLPasses/DpasFuncs/DpasFuncsResolution.cpp#L457, number of elements should be equal to $rc or $sd. However, $rc should be equals to M. so seems like we should not linearize the 2D submatrix.

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td (hunk 2 above)
Comment (whitneywhtsang) [reply mlir/include/mlir/Dialect/LLVMIR/GENXOps.td:387]: https://github.com/pengtu/llvm-project/pull/30

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td (hunk 3 above)
Comment (whitneywhtsang) [reply mlir/include/mlir/Dialect/LLVMIR/GENXOps.td:391]: https://github.com/pengtu/llvm-project/pull/30

**Pull request merged**